tests/unit/spec/newsletter/recovery.js

```diff
@@ -0,0 +1,143 @@
+          <p>Success! An email has been sent to you with your preference center link. Thanks!</p>
+        </div>
+
+        <div class="newsletter-recovery-form-fields">
+          <div class="mzp-c-field mzp-l-stretch">
+            <label for="id_email" class="mzp-c-field-label">Your email address:</label>
+            <input type="email" name="email" required="" class="mzp-c-field-control" id="id_email">
+          </div>
+          <div class="mzp-c-button-container mzp-l-stretch">
+            <button class="mzp-c-button" id="newsletter-submit" type="submit">Send me a link</button>
+          </div>
+        </div>
+      </form>`;
+        document.body.insertAdjacentHTML('beforeend', form);
+    });
+
+    afterEach(function () {
+        const form = document.getElementById('newsletter-recovery-form');
+        form.parentNode.removeChild(form);
+    });
+
+    describe('form submission', function () {
+        let xhr;
+        let xhrRequests = [];
+
+        beforeEach(function () {
+            xhr = sinon.useFakeXMLHttpRequest();
+            xhr.onCreate = (req) => {
+                xhrRequests.push(req);
+            };
+        });
+
+        afterEach(function () {
+            xhr.restore();
+            xhrRequests = [];
+        });
+
+        it('should handle success', function () {
+            spyOn(RecoveryEmailForm, 'handleFormSuccess').and.callThrough();
+            RecoveryEmailForm.init();
+            document.getElementById('id_email').value = 'success@example.com';
+            document.getElementById('newsletter-submit').click();
+            xhrRequests[0].respond(
+                200,
+                { 'Content-Type': 'application/json' },
+                '{"status": "ok"}'
+            );
+
+            expect(RecoveryEmailForm.handleFormSuccess).toHaveBeenCalled();
+            expect(
+                document
+                    .querySelector('.newsletter-recovery-form-success-msg')
+                    .classList.contains('show')
+            ).toBeTrue();
+        });
+
+        it('should handle invalid email', function () {
+            spyOn(RecoveryEmailForm, 'handleFormError').and.callThrough();
+            RecoveryEmailForm.init();
+            document.getElementById('id_email').value = 'invalid@email';
+            document.getElementById('newsletter-submit').click();
+            xhrRequests[0].respond(
+                400,
+                { 'Content-Type': 'application/json' },
+                '{"status": "error", "desc": "Invalid email address"}'
+            );
+
+            expect(RecoveryEmailForm.handleFormError).toHaveBeenCalledWith(
+                'Invalid email address'
+            );
+            expect(
+                document
+                    .querySelector('.error-email-invalid')
+                    .classList.contains('show')
+            ).toBeTrue();
+        });
+
+        it('should handle unknown email', function () {
```

Comment:
**suggestion (non-blocking):** Add test for error with response message other than 'email address not known' (should handle non-user-facing error, shows generic Try Again error message)

Reply:
Good suggestion 👍

Reply:
Oh, looks like I already added an additional test called `should handle failure`, which seems like it should cover this already, unless I'm misunderstanding?

Reply:
I'm looking for another 400 error with desc so it will have a message param but not one that matches the if conditions, so we know it falls through to the else block
` '{"status": "error", "desc": "Not helpful error for user"}'`

Reply:
Ah, ok I understand!